### ISSUE TITLE
Inlined struct copies via params, returns and assignment not elided

### ISSUE THREAD

**Issue by benaadams** [dotnet/runtime#12219]
As seen in `ValueTaskAwaiter` from https://github.com/dotnet/coreclr/pull/22735, https://github.com/dotnet/coreclr/pull/22738 wasn't a specific issue for it with simple repo so opening this.

```csharp
using System.Runtime.CompilerServices;

class Program
{
    static void Main(string[] args)
    {
        InlinedAssignment();
        InlinedCtor();
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static long InlinedAssignment()
    {
        var s = CreateLargeStruct();
        s = GetLargeStruct(s);

        return s.l3;
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static long InlinedCtor()
    {
        var s = new LargeStruct2(CreateLargeStruct());

        return s._largeStruct.l3;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    static LargeStruct GetLargeStruct(LargeStruct l) => l;

    [MethodImpl(MethodImplOptions.NoInlining)]
    static LargeStruct CreateLargeStruct() => new LargeStruct();
}

readonly struct LargeStruct
{
    public readonly long l0;
    public readonly long l1;
    public readonly long l2;
    public readonly long l3;
}

readonly struct LargeStruct2
{
    public readonly LargeStruct _largeStruct;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public LargeStruct2(LargeStruct largeStruct)
    {
        _largeStruct = largeStruct;
    }
}
```
Produces
```asm
; Assembly listing for method Program:InlinedAssignment():long
;  V00 OutArgs      [V00    ] (  1,  1   )  lclBlk (32) [rsp+0x00]   "OutgoingArgSpace"
;  V01 tmp1         [V01    ] (  2,  4   )  struct (32) [rsp+0x48]   do-not-enreg[XSB] addr-exposed "struct address for call/obj"
;* V02 tmp2         [V02    ] (  0,  0   )  struct (32) zero-ref    "struct address for call/obj"
;  V03 tmp3         [V03    ] (  2,  4   )  struct (32) [rsp+0x28]   do-not-enreg[XSB] addr-exposed "Inlining Arg"
;* V04 tmp4         [V04    ] (  0,  0   )    long  ->  zero-ref    V02.l0(offs=0x00) P-INDEP "field V02.l0 (fldOffset=0x0)"
;* V05 tmp5         [V05    ] (  0,  0   )    long  ->  zero-ref    V02.l1(offs=0x08) P-INDEP "field V02.l1 (fldOffset=0x8)"
;* V06 tmp6         [V06    ] (  0,  0   )    long  ->  zero-ref    V02.l2(offs=0x10) P-INDEP "field V02.l2 (fldOffset=0x10)"
;  V07 tmp7         [V07,T01] (  2,  2   )    long  ->  rax         V02.l3(offs=0x18) P-INDEP "field V02.l3 (fldOffset=0x18)"
;  V08 tmp8         [V08,T00] (  3,  6   )   byref  ->  rax         "BlockOp address local"
; Lcl frame size = 104

G_M60975_IG01:
       4883EC68             sub      rsp, 104
       C5F877               vzeroupper 

G_M60975_IG02:
       488D4C2448           lea      rcx, bword ptr [rsp+48H]
       E8BF67FFFF           call     Program:CreateLargeStruct():struct
       C5FA6F442448         vmovdqu  xmm0, qword ptr [rsp+48H]
       C5FA7F442428         vmovdqu  qword ptr [rsp+28H], xmm0
       C5FA6F442458         vmovdqu  xmm0, qword ptr [rsp+58H]
       C5FA7F442438         vmovdqu  qword ptr [rsp+38H], xmm0
       488D442428           lea      rax, bword ptr [rsp+28H]
       488B10               mov      rdx, qword ptr [rax]
       488B4018             mov      rax, qword ptr [rax+24]

G_M60975_IG03:
       4883C468             add      rsp, 104
       C3                   ret      

; Total bytes of code 58, prolog size 7 for method Program:InlinedAssignment():long
```
and
```asm
; Assembly listing for method Program:InlinedCtor():long
;  V00 OutArgs      [V00    ] (  1,  1   )  lclBlk (32) [rsp+0x00]   "OutgoingArgSpace"
;  V01 tmp1         [V01,T00] (  2,  4   )  struct (32) [rsp+0x68]   do-not-enreg[SFB] "NewObj constructor temp"
;  V02 tmp2         [V02    ] (  2,  4   )  struct (32) [rsp+0x48]   do-not-enreg[XSB] addr-exposed "struct address for call/obj"
;  V03 tmp3         [V03,T01] (  2,  4   )  struct (32) [rsp+0x28]   do-not-enreg[SB] "Inlining Arg"
;
; Lcl frame size = 136

G_M31928_IG01:
       4881EC88000000       sub      rsp, 136
       C5F877               vzeroupper 

G_M31928_IG02:
       488D4C2448           lea      rcx, bword ptr [rsp+48H]
       E8BCFFFFFF           call     Program:CreateLargeStruct():struct
       C5FA6F442448         vmovdqu  xmm0, qword ptr [rsp+48H]
       C5FA7F442428         vmovdqu  qword ptr [rsp+28H], xmm0
       C5FA6F442458         vmovdqu  xmm0, qword ptr [rsp+58H]
       C5FA7F442438         vmovdqu  qword ptr [rsp+38H], xmm0
       C5FA6F442428         vmovdqu  xmm0, qword ptr [rsp+28H]
       C5FA7F442468         vmovdqu  qword ptr [rsp+68H], xmm0
       C5FA6F442438         vmovdqu  xmm0, qword ptr [rsp+38H]
       C5FA7F442478         vmovdqu  qword ptr [rsp+78H], xmm0
       488B842480000000     mov      rax, qword ptr [rsp+80H]

G_M31928_IG03:
       4881C488000000       add      rsp, 136
       C3                   ret      

; Total bytes of code 84, prolog size 10 for method Program:InlinedCtor():long
```
Some of these copies could be skipped?

/cc @AndyAyersMS @mikedn @stephentoub @jkotas 

category:cq
theme:structs
skill-level:expert
cost:large

**Comment by kunalspathak** [comment]
Current code generated for x64:

```asm
; InlinedAssignment():long

G_M927_IG01:
       sub      rsp, 104
       vzeroupper
                                                ;; bbWeight=1    PerfScore 1.25
G_M927_IG02:
       lea      rcx, [rsp+48H]
       call     MiniBench.Issue_12219:CreateLargeStruct():MiniBench.LargeStruct
       vmovdqu  xmm0, xmmword ptr [rsp+48H]
       vmovdqu  xmmword ptr [rsp+28H], xmm0
       vmovdqu  xmm0, xmmword ptr [rsp+58H]
       vmovdqu  xmmword ptr [rsp+38H], xmm0
       mov      rax, qword ptr [rsp+40H]
                                                ;; bbWeight=1    PerfScore 10.50
G_M927_IG03:
       add      rsp, 104
       ret
                                                ;; bbWeight=1    PerfScore 1.25
```


```asm
; InlinedCtor():long

G_M47270_IG01:
       sub      rsp, 136
       vzeroupper
                                                ;; bbWeight=1    PerfScore 1.25
G_M47270_IG02:
       lea      rcx, [rsp+48H]
       call     MiniBench.Issue_12219:CreateLargeStruct():MiniBench.LargeStruct
       vmovdqu  xmm0, xmmword ptr [rsp+48H]
       vmovdqu  xmmword ptr [rsp+28H], xmm0
       vmovdqu  xmm0, xmmword ptr [rsp+58H]
       vmovdqu  xmmword ptr [rsp+38H], xmm0
       vmovdqu  xmm0, xmmword ptr [rsp+28H]
       vmovdqu  xmmword ptr [rsp+68H], xmm0
       vmovdqu  xmm0, xmmword ptr [rsp+38H]
       vmovdqu  xmmword ptr [rsp+78H], xmm0
       mov      rax, qword ptr [rsp+80H]
                                                ;; bbWeight=1    PerfScore 18.50
G_M47270_IG03:
       add      rsp, 136
       ret

```

Current code generated for arm64:

```asm
; InlinedAssignment():long

G_M927_IG01:
            stp     fp, lr, [sp,#-80]!
            mov     fp, sp
                                                ;; bbWeight=1    PerfScore 1.50
G_M927_IG02:
            add     x8, fp, #48 // [V01 tmp1]
            bl      MiniBench.Issue_12219:CreateLargeStruct():MiniBench.LargeStruct
            ldp     x0, x1, [fp,#48]    // [V01 tmp1]
            stp     x0, x1, [fp,#16]    // [V03 tmp3]
            ldp     x0, x1, [fp,#64]    // [V01 tmp1+0x10]
            stp     x0, x1, [fp,#32]    // [V03 tmp3+0x10]
            ldr     x0, [fp,#40]        // [V03 tmp3+0x18]
                                                ;; bbWeight=1    PerfScore 11.50
G_M927_IG03:
            ldp     fp, lr, [sp],#80
            ret     lr

```

```asm
; InlinedCtor():long

G_M47270_IG01:
            stp     fp, lr, [sp,#-112]!
            mov     fp, sp
                                                ;; bbWeight=1    PerfScore 1.50
G_M47270_IG02:
            add     x8, fp, #48 // [V02 tmp2]
            bl      MiniBench.Issue_12219:CreateLargeStruct():MiniBench.LargeStruct
            ldp     x0, x1, [fp,#48]    // [V02 tmp2]
            stp     x0, x1, [fp,#16]    // [V03 tmp3]
            ldp     x0, x1, [fp,#64]    // [V02 tmp2+0x10]
            stp     x0, x1, [fp,#32]    // [V03 tmp3+0x10]
            ldp     x0, x1, [fp,#16]    // [V03 tmp3]
            stp     x0, x1, [fp,#80]    // [V01 tmp1]
            ldp     x0, x1, [fp,#32]    // [V03 tmp3+0x10]
            stp     x0, x1, [fp,#96]    // [V01 tmp1+0x10]
            ldr     x0, [fp,#104]       // [V01 tmp1+0x18]
                                                ;; bbWeight=1    PerfScore 19.50
G_M47270_IG03:
            ldp     fp, lr, [sp],#112
            ret     lr

```

Adding the `in` keyword improves the code.

**Comment by tannergooding** [comment]
> Adding the in keyword improves the code.

Could you clarify if this is on Windows or Unix? The ABIs, particularly around SIMD types here isn't the same and so there are scenarios where `in` will likely hurt other perf or where passing by value is expected and so where we'd still want to optimize the byvalue case (particularly given the framework design guideline recommendations around using `in`, which is basically "don't").

**Comment by kunalspathak** [comment]
This is on windows.

**Comment by kunalspathak** [comment]
> Also looks like SqlString:NotEquals(struct,struct):struct might be an interesting case study.

Here is a repro:

```c#
    [MethodImpl(MethodImplOptions.NoInlining)]
    public static bool Compare(SqlString s1, SqlString s2)
    {
        return MyCompare(s1, s2);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool MyCompare(SqlString s1, SqlString s2)
    {
        return MyCompare2(s1, s2);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool MyCompare2(SqlString s1, SqlString s2)
    {
        return s1.IsNull == s2.IsNull;
    }
```

```asm
; Assembly listing for method C:Compare(System.Data.SqlTypes.SqlString,System.Data.SqlTypes.SqlString):bool
; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; optimized code
; rsp based frame
; partially interruptible
; No PGO data
; 0 inlinees with PGO data; 4 single block inlinees; 0 inlinees without PGO data
; Final local variable assignments
;
;  V00 arg0         [V00,T00] (  3,  6   )   byref  ->  rcx         single-def
;  V01 arg1         [V01,T01] (  3,  6   )   byref  ->  rdx         single-def
;# V02 OutArgs      [V02    ] (  1,  1   )  lclBlk ( 0) [rsp+00H]   "OutgoingArgSpace"
;  V03 tmp1         [V03,T02] (  2,  4   )  struct (32) [rsp+68H]   do-not-enreg[S] single-def "Inlining Arg"
;  V04 tmp2         [V04,T03] (  2,  4   )  struct (32) [rsp+48H]   do-not-enreg[S] single-def "Inlining Arg"
;  V05 tmp3         [V05,T04] (  2,  4   )  struct (32) [rsp+28H]   do-not-enreg[SF] ld-addr-op single-def "Inlining Arg"
;  V06 tmp4         [V06,T05] (  2,  4   )  struct (32) [rsp+08H]   do-not-enreg[SF] ld-addr-op single-def "Inlining Arg"
;  V07 tmp5         [V07,T06] (  2,  4   )     int  ->  rax         "impAppendStmt"
;
; Lcl frame size = 136

G_M49149_IG01:
       sub      rsp, 136
       vzeroupper 
						;; bbWeight=1    PerfScore 1.25
G_M49149_IG02:
       vmovdqu  ymm0, ymmword ptr[rcx]
       vmovdqu  ymmword ptr[rsp+68H], ymm0
						;; bbWeight=1    PerfScore 6.00
G_M49149_IG03:
       vmovdqu  ymm0, ymmword ptr[rdx]
       vmovdqu  ymmword ptr[rsp+48H], ymm0
						;; bbWeight=1    PerfScore 6.00
G_M49149_IG04:
       vmovdqu  ymm0, ymmword ptr[rsp+68H]
       vmovdqu  ymmword ptr[rsp+28H], ymm0
						;; bbWeight=1    PerfScore 5.00
G_M49149_IG05:
       vmovdqu  ymm0, ymmword ptr[rsp+48H]
       vmovdqu  ymmword ptr[rsp+08H], ymm0
						;; bbWeight=1    PerfScore 5.00
G_M49149_IG06:
       cmp      byte  ptr [rsp+40H], 0
       sete     al
       movzx    rax, al
       cmp      byte  ptr [rsp+20H], 0
       sete     dl
       movzx    rdx, dl
       cmp      eax, edx
       sete     al
       movzx    rax, al
						;; bbWeight=1    PerfScore 8.00
G_M49149_IG07:
       vzeroupper 
       add      rsp, 136
       ret    
```

**Comment by jakobbotsch** [comment]
I believe this was fixed by #64130. Codegen on main is:

```asm
; Method Program:InlinedAssignment():long
G_M53179_IG01:
       sub      rsp, 72
						;; size=4 bbWeight=1    PerfScore 0.25

G_M53179_IG02:
       lea      rcx, [rsp+28H]
       call     [Program:CreateLargeStruct():LargeStruct]
       mov      rax, qword ptr [rsp+40H]
						;; size=16 bbWeight=1    PerfScore 4.50

G_M53179_IG03:
       add      rsp, 72
       ret      
						;; size=5 bbWeight=1    PerfScore 1.25
; Total bytes of code: 25
```

```asm
; Method Program:InlinedCtor():long
G_M38338_IG01:
       sub      rsp, 72
						;; size=4 bbWeight=1    PerfScore 0.25

G_M38338_IG02:
       lea      rcx, [rsp+28H]
       call     [Program:CreateLargeStruct():LargeStruct]
       mov      rax, qword ptr [rsp+40H]
						;; size=16 bbWeight=1    PerfScore 4.50

G_M38338_IG03:
       add      rsp, 72
       ret      
						;; size=5 bbWeight=1    PerfScore 1.25
; Total bytes of code: 25
```
